src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1528,13 +1539,20 @@ size_t ObjectSynchronizer::deflate_idle_monitors(ObjectMonitorsHashtable* table)
       // ObjectMonitors that were deflated in this cycle.
       HandshakeForDeflation hfd_hc;
       Handshake::execute(&hfd_hc);
+      // Also, we sync and desync GC threads around the handshake, so that they can
+      // safely read the mark-word and look-through to the object-monitor, without
+      // being afraid that the object-monitor is going away.
+      VM_RendezvousGCThreads sync_gc;
+      VMThread::execute(&sync_gc);
 
       if (ls != nullptr) {
         ls->print_cr("after handshaking: in_use_list stats: ceiling="
                      SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
                      in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
         timer.start();
       }
+    } else {
+      assert(current->is_VM_thread(), "only VM thread may get here");
```

Comment:
Actually, maybe we should assert we are at safepoint? To protect the accidental call from the VMOp that is executed by VMThread, but outside of safepoint.

```
} else {
  // This is not a monitor deflation thread.
  // No handshake or rendezvous is needed when we are already at safepoint.
  assert_at_safepoint();
}
```